W-14403459-governance-performance-issues-fa (#217)
mulesoft/docs-design-center: modules/ROOT/pages/design-add-api-dependency.adoc

When you have your branch, a pop-up appears where you can see the missing ruleset filtered by tags, category, and lifecycle state. To edit those filters, click *Edit criteria*. When you are done, click *Add all rulesets*. +
_Result_: The rulesets are added to your project.

== Improving Ruleset Validation

You can validate your rulesets manually or automatically:

> * To turn manual validation on, click *Validate for Governance* in the *Governance* tab of the *Project Errors* section of the text editor. 

Reviewer: Actually, you need to enable manual mode first by clicking on the Green auto-validation button toggle to switch it off. Then, the "Validate for Governance" button is enabled and user can trigger validation when clicking the button.

* To enable auto-validation again, click *Governance auto-validation* in the *Governance* tab of the *Project Errors* section of the text editor. If you click *Governance auto-validation* again, it will disable it. 

> * If the validation process takes too long due to the complexity of either the rulesets added or the API specification, a warning message appears informing auto-validation is disabled for that project. CLick *Ok* to change to manual mode.

Reviewer 2: Click has capital L incorrectly.

_Result_: The rulesets are added to your project.

[[fix-gov-issues]]
== Fix Governance Conformance Issues

> When adding the rulesets as dependencies, you can view the conformance messages in the *Project Errors* section of the text editor. Use the provided information to fix any conformance issues and then republish your API specification. 

Reviewer 2: Shouldn't we say `governance validations` instead of `conformance messages`?

Reviewer 3: @ignaciovazgueza, we should be consistent with the API Governance terminology, as this is part of the API Governance user journey. APIs are either conformant or nonconformant to rulesets.

Reviewer 3: Also, @failup, I noticed you changed this to "When adding" instead of "After you add" the rulesets as dependencies. The dependencies have to actually be added to see the errors appear in the Project Errors page, right, or has the timing on that changed? My understanding is that the rulesets are added when they click the plus icon and that dependency is auto-saved in APID. Are you interpreting this otherwise? If so, we should probably explain more here.

Reviewer 3: Looking closer at this, I think the issue is that the validation is put in the wrong section. This either needs to be in its own section above Fix Governance Conformance Issues, or it needs to be integrated more fully into the section and the name of the section changed to encompass both running the validation and fixing the conformance.

Reviewer 3: Also, as we discussed in our sync last week, I suggest reorganizing this new content so that it has a simpler lead-in and has the "what to do" in a procedure rather than spread out into paragraphs.
The validation is done automatically by default. If you get a message that the validation is taking too long:
1) Make the following changes to improve performance:
 (list things to do to avoid the situation here instead of in a troubleshooting section)
2) Click *Validate for Governance*.
...

Reviewer 3: Is there any point of them retrying (clicking Ok when they get the error) before taking steps to improve the performance? If so, you could put that as the first step in the procedure followed by the ones I mention above (the steps are paraphrased to show organization suggestions, not specific text suggestions).

> If the validation process takes too long due to the complexity of the rulesets added, a warning message appears informing auto-validation is disabled for that project. CLick *Ok* to change to manual mode.

Reviewer 2: ```suggestion
If the validation process takes too long, a warning message will appear informing governance auto-validation is going to be disabled for that project. Click *Ok* to change to manual mode.
```

Author: we don't use future in our docs!

> If the validation process takes too long due to the complexity of the rulesets added, a warning message appears informing auto-validation is disabled for that project. CLick *Ok* to change to manual mode.

Reviewer: "due to the complexity of either the added rulesets or the API specification"

> If the validation process takes too long due to the complexity of the rulesets added, a warning message appears informing auto-validation is disabled for that project. CLick *Ok* to change to manual mode.

Reviewer 3: ```suggestion
If the automatic validation process takes too long, you get a warning that auto-validation is disabled for the project. Click *Ok* to change to manual mode.
```

If you don't want to disable auto-validation, click *Cancel*. 

> NOTE: If the auto-validation of rulesets still takes too long, the warning message will re-appear. 

Reviewer 3: ```suggestion
NOTE: If the auto-validation of rulesets still takes too long, you get the warning message again.
```

=== Validating Rulesets Manually

> You can validate the rulesets manually. To do so, click *Validate for Governance* in the *Governance* tab of the *Project Errors* section of the text editor.

Reviewer 2: ```suggestion
You can now validate the rulesets manually. To do so, click *Validate for Governance* icon button in the *Governance* tab of the *Project Errors* section of the text editor.
```

Author: @ignaciovazqueza we have certain rules to follow when it comes to writing, these are the standards we are currently applying, we can't say icon button. Also, since this is a docs page and not a release not document we shouldn't say "now"!

> You can also enable auto-validation again. To do it, click *Governance auto-validation* in the *Governance* tab of the *Project Errors* section of the text editor. If you click *Governance auto-validation* again, it will disable it.

Reviewer 2: ```suggestion
You can also enable auto-validation again. To do it, click *Governance auto-validation* icon button in the *Governance* tab of the *Project Errors* section of the text editor. If you click *Governance auto-validation* again, it will disable it.
```

=== Troubleshooting 

The ruleset validation process can experience delays due to the following issues:

* *Browser Resources*: API Designer needs browser resources to validate rulesets. Close unnecessary tabs or applications to free up resources, and ensure your device is running efficiently and has enough memory and CPU resources.

> * *Rulesets*: As additional rulesets are added as dependencies to a project, validation times inevitably increases due to the number of diagnostics being computed. Switch to manual validation to reduce validation times.

Reviewer 2: Switching to manual validation won't reduce validation times. It will remove our 30sec timeout so if validation exceeds that timeout, it will even take longer than auto-validation

Reviewer: I would phrase this as "Switch to manual validation to prevent from triggering multiple automatic validations that may exhaust your available resources. Instead, manual mode will trigger a single validation process per ruleset."

> * *API Complexity*: If your API is modularized, has complex recursive types or circular references, the validation process takes more time. Switch to manual validation to reduce validation times.

Reviewer 2: Same comment here, manual mode doesn't reduce validation times.

Reviewer: The main goal of switching to manual validation is to prevent multiple validations being triggered automatically when editing the API spec. With manual mode, the user decides when to trigger the resource intensive validation
